README.md
ensured_bufreader
=======

[![ensured_bufreader at crates.io](https://img.shields.io/crates/v/ensured_bufreader.svg)](https://crates.io/crates/ensured_bufreader)
[![ensured_bufreader at docs.rs](https://docs.rs/ensured_bufreader/badge.svg)](https://docs.rs/ensured_bufreader)
[![Build Status (master)](https://travis-ci.org/IgaguriMK/ensured_bufreader.svg?branch=master)](https://travis-ci.org/IgaguriMK/ensured_bufreader)
[![Build Status (dev)](https://travis-ci.org/IgaguriMK/ensured_bufreader.svg?branch=dev)](https://travis-ci.org/IgaguriMK/ensured_bufreader)

## Overview

`ensured_bufreader` provides `EnsuredBufReader` that _ensured_ length bytes in its buffer if it can read from underlying reader.

`std::io::BufReader` doesn't read bytes from underlying reader if it has buffered bytes.
This behavior is better if you need buffering for performance.
But, if you need buffering for algorithm such as peeking N bytes, `BufReader` donesn't ensure N bytes in its buffer.

If there are too few bytes in buffer when `.fill_buf()` called, `EnsuredBufReader` tries to read additional bytes from underlying reader and keep `.fill_buf()?.len() > ensured`.

## Comparition with other crates

`buffered-reader` provides same functionality as this crate.
But there is some differences.

|   | `ensured_bufreader` | `buffered-reader = "0.10.0"` |
|:-:|:--|:--|
| Implementation strategy | Uses standard trait `BufRead` and simple wrapper struct | Provides new trait and some implementaions |
| When returns error | Immediately | Saved and returned when read position error occuered |
| License | MIT or Apache-2.0 (permissive) | GPL-3.0 (copyleft) |

## License

`ensured_bufreader` is distributed under the terms of both the MIT license and the Apache License (Version 2.0).

See [LICENSE-APACHE](LICENSE-APACHE) and [LICENSE-MIT](LICENSE-MIT).

CHANGELOG.md
## Version 0.2.0

### Breaking Changes

* Change
    - `EnsuredBufReader` now takes buffer type parameter `B`.
* Removed
    - `EnsuredBufReader::with_capacity`
    - `EnsuredBufReader::with_ensured_size`
    - `EnsuredBufReader::fill_buf_to_expected_size`

### New Features

* Now, `EnsuredBufReader` can works with external buffer. For example, you can provide `&mut [u8]` as buffer.
    - From `&mut [u8]`
        + `EnsuredBufReader::from_mut_ref`
        + `EnsuredBufReader::from_mut_ref_and_ensured_size`
    - From any type `B: AsRef<[u8]> + AsMut<[u8]>`
        + `EnsuredBufReader::from_buffer`
        + `EnsuredBufReader::from_buffer_and_ensured_size`
* Instead of `fill_buf_to_expected_size`, new method `fill_buf_to_expected_size` is added.
    - This method returns error when expected size is too large.

## Version 0.1.0

### New Features

* `EnsuredBufReader`, a buffer supports ensures size of buffer contents.

tests/api_tests.rs
use std::io::{BufRead, ErrorKind, Read};

use ensured_bufreader::{EnsuredBufReader, DEFAULT_ENSURED_BYTES};

#[test]
#[should_panic]
fn capacity_is_smaller_than_ensure_not_allowed() {
    let r: &[u8] = &[];
    let _ = EnsuredBufReader::with_capacity_and_ensured_size(100, 101, r);
}

#[test]
#[should_panic]
fn ensure_is_0_not_allowed_with_capacity_and_ensure() {
    let r: &[u8] = &[];
    let _ = EnsuredBufReader::with_capacity_and_ensured_size(1024, 0, r);
}

#[test]
fn read_long() {
    let short = "aÀあ\u{1F600}".as_bytes();
    let mut input = Vec::with_capacity(short.len() * 32 * 1024);
    for _ in 0..32 * 1024 {
        input.extend_from_slice(short);
    }

    let mut r = EnsuredBufReader::new(input.as_slice());

    let mut read_bytes = Vec::<u8>::with_capacity(input.len());
    let mut read_buf = [0u8; 256];

    loop {
        let n = r.read(&mut read_buf).unwrap();

        let bs = &read_buf[..n];
        read_bytes.extend_from_slice(bs);

        if n < DEFAULT_ENSURED_BYTES {
            break;
        }
    }

    assert_eq!(
        input,
        read_bytes,
        "with capacity ={}, ensure = {}",
        r.get_capacity(),
        r.get_ensured_size()
    );
}

#[test]
fn fill_buf_short() {
    let input = "aÀあ\u{1F600}".as_bytes();

    let mut r = EnsuredBufReader::new(input);
    let mut read_bytes = Vec::<u8>::with_capacity(input.len());

    loop {
        let buf = r.fill_buf().unwrap();
        read_bytes.extend_from_slice(buf);
        let n = buf.len();
        r.consume(n);

        if n < DEFAULT_ENSURED_BYTES {
            break;
        }
    }

    assert_eq!(
        input,
        read_bytes.as_slice(),
        "with capacity ={}, ensure = {}",
        r.get_capacity(),
        r.get_ensured_size()
    );
}

#[test]
fn fill_buf_long() {
    let short = "aÀあ\u{1F600}".as_bytes();
    let mut input = Vec::with_capacity(short.len() * 32 * 1024);
    for _ in 0..32 * 1024 {
        input.extend_from_slice(short);
    }

    let mut r = EnsuredBufReader::new(input.as_slice());

    let mut read_bytes = Vec::<u8>::with_capacity(input.len());

    loop {
        let buf = r.fill_buf().unwrap();
        read_bytes.extend_from_slice(buf);
        let n = buf.len();
        r.consume(n);

        if n < DEFAULT_ENSURED_BYTES {
            break;
        }
    }

    assert_eq!(
        input,
        read_bytes,
        "with capacity ={}, ensure = {}",
        r.get_capacity(),
        r.get_ensured_size()
    );
}

#[test]
fn works_with_given_buffer() {
    let short = "aÀあ\u{1F600}".as_bytes();
    let mut input = Vec::with_capacity(short.len() * 32 * 1024);
    for _ in 0..32 * 1024 {
        input.extend_from_slice(short);
    }

    let mut buf = [0u8; 2 * DEFAULT_ENSURED_BYTES];
    let mut r = EnsuredBufReader::from_mut_ref(&mut buf, input.as_slice());

    let mut read_bytes = Vec::<u8>::with_capacity(input.len());

    loop {
        let buf = r.fill_buf().unwrap();
        read_bytes.extend_from_slice(buf);
        let n = buf.len();
        r.consume(n);

        if n < DEFAULT_ENSURED_BYTES {
            break;
        }
    }

    assert_eq!(
        input,
        read_bytes,
        "with capacity ={}, ensure = {}",
        r.get_capacity(),
        r.get_ensured_size()
    );
}

#[test]
fn fill_buf_to_expected_size_returns_error_when_expected_size_is_too_large() {
    let short = "aÀあ\u{1F600}".as_bytes();
    let mut input = Vec::with_capacity(short.len() * 32 * 1024);
    for _ in 0..256 {
        input.extend_from_slice(short);
    }

    let mut r = EnsuredBufReader::with_capacity_and_ensured_size(512, 1, input.as_slice());

    let err = r.fill_buf_to_expected_size(513).unwrap_err();
    assert_eq!(err.kind(), ErrorKind::InvalidInput);
}


tests/integration_test.rs
use std::io::{BufRead, Read};

use ensured_bufreader::EnsuredBufReader;

#[test]
#[should_panic]
fn capacity_is_smaller_than_ensure_not_allowed() {
    let r: &[u8] = &[];
    let _ = EnsuredBufReader::with_capacity_and_ensure(100, 101, r);
}

#[test]
#[should_panic]
fn ensure_is_0_not_allowed_with_ensure() {
    let r: &[u8] = &[];
    let _ = EnsuredBufReader::with_ensure(0, r);
}

#[test]
#[should_panic]
fn ensure_is_0_not_allowed_with_capacity_and_ensure() {
    let r: &[u8] = &[];
    let _ = EnsuredBufReader::with_capacity_and_ensure(1024, 0, r);
}

#[test]
fn read_short() {
    let orig_bytes = "aÀあ\u{1F600}".as_bytes();

    for mut r in util::gen_readers(&orig_bytes) {
        let mut read_bytes = Vec::<u8>::with_capacity(orig_bytes.len());
        let mut read_buf = [0u8; 256];
        loop {
            let n = r.read(&mut read_buf).unwrap();
            eprintln!("{}", n);
            if n == 0 {
                break;
            }
            let bs = &read_buf[..n];
            read_bytes.extend_from_slice(bs);
        }
        assert_eq!(
            orig_bytes,
            read_bytes.as_slice(),
            "with capacity ={}, ensure = {}",
            r.get_capacity(),
            r.get_ensure()
        );
    }
}

#[test]
fn read_long() {
    let short = "aÀあ\u{1F600}".as_bytes();
    let mut orig_bytes = Vec::with_capacity(short.len() * 32 * 1024);
    for _ in 0..32 * 1024 {
        orig_bytes.extend_from_slice(short);
    }

    for mut r in util::gen_readers(&orig_bytes) {
        let mut read_bytes = Vec::<u8>::with_capacity(orig_bytes.len());
        let mut read_buf = [0u8; 256];
        loop {
            let n = r.read(&mut read_buf).unwrap();
            if n == 0 {
                break;
            }
            let bs = &read_buf[..n];
            read_bytes.extend_from_slice(bs);
        }
        assert_eq!(
            orig_bytes,
            read_bytes,
            "with capacity ={}, ensure = {}",
            r.get_capacity(),
            r.get_ensure()
        );
    }
}

#[test]
fn read_expected() {
    let sizes: &[usize] = &[
        1usize,
        2,
        3,
        511,
        512,
        513,
        16 * 1024 - 1,
        16 * 1024,
        16 * 1024 + 1,
        256 * 1024,
    ];

    for &size in sizes {
        let mut orig_bytes = Vec::with_capacity(size);
        orig_bytes.resize(size - 1, 0);
        orig_bytes.push(1);
        assert_eq!(orig_bytes.len(), size);

        for mut r in util::gen_readers(&orig_bytes) {
            let ensure = r.get_ensure();
            loop {
                let buf = r.fill_buf().unwrap();
                let n = buf.len();
                if n == 0 {
                    // EOF
                    break;
                }
                match buf.last() {
                    Some(1) => {} // reach end.
                    _ => assert!(
                        buf.len() >= ensure,
                        "buf.len() (= {}) should be larger than ensure (= {})",
                        buf.len(),
                        ensure
                    ),
                }
                r.consume(n);
            }
        }
    }
}

mod util {
    use ensured_bufreader::EnsuredBufReader;

    pub fn gen_readers<'a>(
        orig_bytes: &'a [u8],
    ) -> impl Iterator<Item = EnsuredBufReader<&'a [u8]>> {
        vec![
            EnsuredBufReader::new(orig_bytes),
            EnsuredBufReader::with_capacity(1, orig_bytes),
            EnsuredBufReader::with_capacity(512, orig_bytes),
            EnsuredBufReader::with_ensure(1, orig_bytes),
            EnsuredBufReader::with_ensure(16 * 1024, orig_bytes),
            EnsuredBufReader::with_capacity_and_ensure(1, 1, orig_bytes),
            EnsuredBufReader::with_capacity_and_ensure(3, 3, orig_bytes),
        ]
        .into_iter()
    }
}


tests/integration_tests.rs
use std::io::{BufRead, Read};

use ensured_bufreader::EnsuredBufReader;

#[test]
#[should_panic]
fn capacity_is_smaller_than_ensure_not_allowed() {
    let r: &[u8] = &[];
    let _ = EnsuredBufReader::with_capacity_and_ensured_size(100, 101, r);
}

#[test]
#[should_panic]
fn ensure_is_0_not_allowed_with_ensure() {
    let r: &[u8] = &[];
    let _ = EnsuredBufReader::with_ensured_size(0, r);
}

#[test]
#[should_panic]
fn ensure_is_0_not_allowed_with_capacity_and_ensure() {
    let r: &[u8] = &[];
    let _ = EnsuredBufReader::with_capacity_and_ensured_size(1024, 0, r);
}

#[test]
fn read_short() {
    let orig_bytes = "aÀあ\u{1F600}".as_bytes();

    for mut r in util::gen_readers(&orig_bytes) {
        let mut read_bytes = Vec::<u8>::with_capacity(orig_bytes.len());
        let mut read_buf = [0u8; 256];
        loop {
            let n = r.read(&mut read_buf).unwrap();
            eprintln!("{}", n);
            if n == 0 {
                break;
            }
            let bs = &read_buf[..n];
            read_bytes.extend_from_slice(bs);
        }
        assert_eq!(
            orig_bytes,
            read_bytes.as_slice(),
            "with capacity ={}, ensure = {}",
            r.get_capacity(),
            r.get_ensured_size()
        );
    }
}

#[test]
fn read_long() {
    let short = "aÀあ\u{1F600}".as_bytes();
    let mut orig_bytes = Vec::with_capacity(short.len() * 32 * 1024);
    for _ in 0..32 * 1024 {
        orig_bytes.extend_from_slice(short);
    }

    for mut r in util::gen_readers(&orig_bytes) {
        let mut read_bytes = Vec::<u8>::with_capacity(orig_bytes.len());
        let mut read_buf = [0u8; 256];
        loop {
            let n = r.read(&mut read_buf).unwrap();
            if n == 0 {
                break;
            }
            let bs = &read_buf[..n];
            read_bytes.extend_from_slice(bs);
        }
        assert_eq!(
            orig_bytes,
            read_bytes,
            "with capacity ={}, ensure = {}",
            r.get_capacity(),
            r.get_ensured_size()
        );
    }
}

#[test]
fn read_expected() {
    let sizes: &[usize] = &[
        1usize,
        2,
        3,
        511,
        512,
        513,
        16 * 1024 - 1,
        16 * 1024,
        16 * 1024 + 1,
        256 * 1024,
    ];

    for &size in sizes {
        let mut orig_bytes = Vec::with_capacity(size);
        orig_bytes.resize(size - 1, 0);
        orig_bytes.push(1);
        assert_eq!(orig_bytes.len(), size);

        for mut r in util::gen_readers(&orig_bytes) {
            let ensure = r.get_ensured_size();
            loop {
                let buf = r.fill_buf().unwrap();
                let n = buf.len();
                if n == 0 {
                    // EOF
                    break;
                }
                match buf.last() {
                    Some(1) => {} // reach end.
                    _ => assert!(
                        buf.len() >= ensure,
                        "buf.len() (= {}) should be larger than ensure (= {})",
                        buf.len(),
                        ensure
                    ),
                }
                r.consume(n);
            }
        }
    }
}

mod util {
    use ensured_bufreader::EnsuredBufReader;

    pub fn gen_readers<'a>(
        orig_bytes: &'a [u8],
    ) -> impl Iterator<Item = EnsuredBufReader<&'a [u8]>> {
        vec![
            EnsuredBufReader::new(orig_bytes),
            EnsuredBufReader::with_capacity(1, orig_bytes),
            EnsuredBufReader::with_capacity(512, orig_bytes),
            EnsuredBufReader::with_ensured_size(1, orig_bytes),
            EnsuredBufReader::with_ensured_size(16 * 1024, orig_bytes),
            EnsuredBufReader::with_capacity_and_ensured_size(1, 1, orig_bytes),
            EnsuredBufReader::with_capacity_and_ensured_size(3, 3, orig_bytes),
        ]
        .into_iter()
    }
}


src/lib.rs
//! Provides `EnsuredBuffer` that impls [`BufRead`](https://doc.rust-lang.org/std/io/trait.BufRead.html).
//!
//!

#![warn(missing_docs)]
use std::error;
use std::fmt;
use std::io::{self, BufRead, Read};

/// Default buffer _capacity_
///
/// Current value is 8 kiB, but may change in the future.
///
/// # Examples
///
/// ```
/// use std::fs::File;
/// use ensured_bufreader::{DEFAULT_BUFFER_SIZE, EnsuredBufReader};
///
/// fn main() -> std::io::Result<()> {
///     let f = File::open("README.md")?;
///     let r = EnsuredBufReader::new(f);
///
///     assert_eq!(r.get_capacity(), DEFAULT_BUFFER_SIZE);
///     Ok(())
/// }
/// ```
pub const DEFAULT_BUFFER_SIZE: usize = 8 * 1024;

/// Default _ensured_ size.
///
/// Current value is 128 B, but may change in the future.
/// # Examples
///
/// ```
/// use std::fs::File;
/// use ensured_bufreader::{DEFAULT_ENSURED_BYTES, EnsuredBufReader};
///
/// fn main() -> std::io::Result<()> {
///     let f = File::open("README.md")?;
///     let r = EnsuredBufReader::new(f);
///
///     assert_eq!(r.get_ensured_size(), DEFAULT_ENSURED_BYTES);
///     Ok(())
/// }
/// ```
pub const DEFAULT_ENSURED_BYTES: usize = 128;

/// A [`BufRead`](https://doc.rust-lang.org/std/io/trait.BufRead.html)er that ensures _ensured_ bytes in buffer.
///
/// `EnsuredBufReader` keeps _ensured_ bytes in buffer if it can read from underlying reader.
/// To fetch bytes into buffer, call `fill_buf()`.
pub struct EnsuredBufReader<R, B>
where
    R: Read,
    B: AsRef<[u8]> + AsMut<[u8]>,
{
    inner: R,
    buf: B,
    pos: usize,
    cap: usize,
    ensured_size: usize,
}

impl<R: Read> EnsuredBufReader<R, Vec<u8>> {
    /// Creates a new `EnsuredBufReader` with a default _capacity_ (`DEFAULT_BUFFER_SIZE`) and a default _ensured_ size (`DEFAULT_ENSURED_BYTES`).
    ///
    /// # Examples
    ///
    /// ```
    /// use std::fs::File;
    /// use ensured_bufreader::EnsuredBufReader;
    ///
    /// fn main() -> std::io::Result<()> {
    ///     let f = File::open("README.md")?;
    ///     let r = EnsuredBufReader::new(f);
    ///     Ok(())
    /// }
    /// ```
    pub fn new(inner: R) -> EnsuredBufReader<R, Vec<u8>> {
        EnsuredBufReader::with_capacity_and_ensured_size(
            DEFAULT_BUFFER_SIZE,
            DEFAULT_ENSURED_BYTES,
            inner,
        )
    }

    /// Creates a new `EnsuredBufReader` with a specified `capacity` and `ensured_size`.
    ///
    /// `capacity` must be larger than or equal to `ensured_size`.
    /// `ensured_size` must be positive.
    ///
    /// # Panics
    ///
    /// Panics if `capacity` is smaller than `ensured_size`.
    /// Panics if `ensured_size` is 0.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::fs::File;
    /// use ensured_bufreader::EnsuredBufReader;
    ///
    /// fn main() -> std::io::Result<()> {
    ///     let f = File::open("README.md")?;
    ///     let r = EnsuredBufReader::with_capacity_and_ensured_size(1024, 32, f);
    ///     Ok(())
    /// }
    /// ```
    pub fn with_capacity_and_ensured_size(
        capacity: usize,
        ensured_size: usize,
        inner: R,
    ) -> EnsuredBufReader<R, Vec<u8>> {
        assert_ne!(ensured_size, 0, "'ensure' must be positive.");
        assert!(
            capacity >= ensured_size,
            "'capacity' ({}) must be larger than or equal to 'ensured_size' ({}).",
            capacity,
            ensured_size
        );
        EnsuredBufReader {
            inner,
            buf: vec![0; capacity],
            pos: 0,
            cap: 0,
            ensured_size,
        }
    }
}

impl<R: Read> EnsuredBufReader<R, &mut [u8]> {
    /// Creates a new `EnsuredBufReader` with given buffer.
    ///
    /// Buffer length must be larger than or equal to [`DEFAULT_ENSURED_BYTES`](constant.DEFAULT_ENSURED_BYTES.html).
    ///
    /// # Panics
    ///
    /// Panics if buffer is smaller than DEFAULT_ENSURED_BYTES.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::fs::File;
    /// use ensured_bufreader::EnsuredBufReader;
    ///
    /// fn main() -> std::io::Result<()> {
    ///     let f = File::open("README.md")?;
    ///     let mut buf = [0u8; 1024];
    ///     let r = EnsuredBufReader::from_mut_ref(&mut buf, f);
    ///     Ok(())
    /// }
    /// ```
    pub fn from_mut_ref(buf: &mut [u8], inner: R) -> EnsuredBufReader<R, &mut [u8]> {
        assert!(
            buf.len() >= DEFAULT_ENSURED_BYTES,
            "buffer size ({}) must be larger than or equal to default ensured size' ({}).",
            buf.len(),
            DEFAULT_ENSURED_BYTES
        );
        EnsuredBufReader::from_mut_ref_and_ensured_size(buf, DEFAULT_ENSURED_BYTES, inner)
    }

    /// Creates a new `EnsuredBufReader` with given buffer and a specified `ensured_size`.
    ///
    /// Buffer length must be larger than or equal to `ensured_size`.
    /// `ensured_size` must be positive.
    ///
    /// # Panics
    ///
    /// Panics if buffer is smaller than `ensured_size`.
    /// Panics if `ensured_size` is 0.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::fs::File;
    /// use ensured_bufreader::EnsuredBufReader;
    ///
    /// fn main() -> std::io::Result<()> {
    ///     let f = File::open("README.md")?;
    ///     let mut buf = [0u8; 1024];
    ///     let r = EnsuredBufReader::from_mut_ref_and_ensured_size(&mut buf, 32, f);
    ///     Ok(())
    /// }
    /// ```
    pub fn from_mut_ref_and_ensured_size(
        buf: &mut [u8],
        ensured_size: usize,
        inner: R,
    ) -> EnsuredBufReader<R, &mut [u8]> {
        assert_ne!(ensured_size, 0, "'ensure' must be positive.");
        assert!(
            buf.len() >= ensured_size,
            "buffer size ({}) must be larger than or equal to 'ensured_size' ({}).",
            buf.len(),
            ensured_size
        );
        EnsuredBufReader {
            inner,
            buf,
            pos: 0,
            cap: 0,
            ensured_size,
        }
    }
}

impl<R: Read, B: AsRef<[u8]> + AsMut<[u8]>> EnsuredBufReader<R, B> {
    /// Creates a new `EnsuredBufReader` with given buffer.
    ///
    /// Buffer length must be larger than or equal to [`DEFAULT_ENSURED_BYTES`](constant.DEFAULT_ENSURED_BYTES.html).
    ///
    /// # Panics
    ///
    /// Panics if buffer is smaller than DEFAULT_ENSURED_BYTES.
    pub fn from_buffer(buf: B, inner: R) -> EnsuredBufReader<R, B> {
        assert!(
            buf.as_ref().len() >= DEFAULT_ENSURED_BYTES,
            "buffer size ({}) must be larger than or equal to 'ensured_size' ({}).",
            buf.as_ref().len(),
            DEFAULT_ENSURED_BYTES
        );
        EnsuredBufReader::from_buffer_and_ensured_size(buf, DEFAULT_ENSURED_BYTES, inner)
    }

    /// Creates a new `EnsuredBufReader` with given buffer and a specified `ensured_size`.
    ///
    /// Buffer length must be larger than or equal to `ensured_size`.
    /// `ensured_size` must be positive.
    ///
    /// # Panics
    ///
    /// Panics if buffer is smaller than `ensured_size`.
    /// Panics if `ensured_size` is 0.
    pub fn from_buffer_and_ensured_size(
        buf: B,
        ensured_size: usize,
        inner: R,
    ) -> EnsuredBufReader<R, B> {
        assert_ne!(ensured_size, 0, "'ensure' must be positive.");
        assert!(
            buf.as_ref().len() >= ensured_size,
            "buffer size ({}) must be larger than or equal to 'ensured_size' ({}).",
            buf.as_ref().len(),
            ensured_size
        );
        EnsuredBufReader {
            inner,
            buf,
            pos: 0,
            cap: 0,
            ensured_size,
        }
    }

    /// Returns a reference to current buffer.
    /// This method doesn't read bytes from underlying reader.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::fs::File;
    /// use std::io::{self, BufRead};
    /// use ensured_bufreader::EnsuredBufReader;
    ///
    /// fn main() -> io::Result<()> {
    ///     let f = File::open("README.md")?;
    ///     let mut r = EnsuredBufReader::new(f);
    ///
    ///     // Read bytes from file and consume 8 bytes.
    ///     let read_bytes = r.fill_buf()?.to_owned();
    ///     r.consume(8);
    ///     
    ///     // Get buffer.
    ///     // Current buffer should be　8 bytes shorter than `read_bytes`.
    ///     let buf = r.buffer();
    ///     assert_eq!(buf, &read_bytes[8..]);
    ///
    ///     Ok(())
    /// }
    /// ```
    pub fn buffer(&self) -> &[u8] {
        &self.buf.as_ref()[self.pos..self.cap]
    }

    /// Try to fill buffer and return reference to buffer.
    /// The buffer filled at least `expected_size` bytes if `EnsuredBufReader` could read from underlying reader.
    ///
    /// # Errors
    ///
    /// Returns error that has `.kind() == ErrorKind::InvalidInput` if `expected_size` is larger than _capacity_.
    ///
    /// # Examples
    ///
    /// The buffer will be filled to `expected_size`.
    ///
    /// ```
    /// use std::fs::File;
    /// use std::io::{self, BufRead};
    /// use ensured_bufreader::EnsuredBufReader;
    ///
    /// fn main() -> io::Result<()> {
    ///     let f = File::open("README.md")?;
    ///     let mut r = EnsuredBufReader::with_capacity_and_ensured_size(1024, 1, f);
    ///
    ///     // Fill buffer.
    ///     let read_bytes = r.fill_buf_to_expected_size(512)?;
    ///     assert!(read_bytes.len() >= 512);
    ///
    ///     Ok(())
    /// }
    /// ```
    ///
    /// If `expected_size` is larger than _capacity_, error will be returned.
    ///
    /// ```
    /// use std::fs::File;
    /// use std::io::{self, BufRead, ErrorKind};
    /// use ensured_bufreader::EnsuredBufReader;
    ///
    /// fn main() -> io::Result<()> {
    ///     let f = File::open("README.md")?;
    ///     let mut r = EnsuredBufReader::with_capacity_and_ensured_size(512, 1, f);
    ///
    ///     let err = r.fill_buf_to_expected_size(513).unwrap_err();
    ///     assert_eq!(err.kind(), ErrorKind::InvalidInput);
    ///
    ///     Ok(())
    /// }
    /// ```
    pub fn fill_buf_to_expected_size(&mut self, expected_size: usize) -> io::Result<&[u8]> {
        if self.current_bytes() >= expected_size {
            return Ok(self.buffer());
        }

        if self.buf.as_mut().len() < expected_size {
            return Err(io::Error::new(
                io::ErrorKind::InvalidInput,
                ExpectedSizeTooLargeError(),
            ));
        }
        if self.buf.as_mut().len() - self.pos < expected_size {
            self.move_buf_to_head()
        }
        while self.current_bytes() < expected_size {
            let n = self.inner.read(&mut self.buf.as_mut()[self.cap..])?;
            if n == 0 {
                // Reach EOF
                break;
            }
            self.cap += n;
        }

        Ok(self.buffer())
    }

    /// Get current _capacity_ size.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::fs::File;
    /// use ensured_bufreader::EnsuredBufReader;
    ///
    /// fn main() -> std::io::Result<()> {
    ///     let f = File::open("README.md")?;
    ///     let r = EnsuredBufReader::new(f);
    ///
    ///     assert_eq!(r.get_capacity(), 8192);
    ///     Ok(())
    /// }
    /// ```
    pub fn get_capacity(&self) -> usize {
        self.buf.as_ref().len()
    }

    /// Get current _ensured_ size.
    ///
    /// # Examples
    ///
    /// ```
    /// use std::fs::File;
    /// use ensured_bufreader::EnsuredBufReader;
    ///
    /// fn main() -> std::io::Result<()> {
    ///     let f = File::open("README.md")?;
    ///     let r = EnsuredBufReader::new(f);
    ///
    ///     assert_eq!(r.get_ensured_size(), 128);
    ///     Ok(())
    /// }
    /// ```
    pub fn get_ensured_size(&self) -> usize {
        self.ensured_size
    }

    /// Returns count of bytes in buffer.
    pub fn current_bytes(&self) -> usize {
        self.cap - self.pos
    }

    fn move_buf_to_head(&mut self) {
        if self.pos == self.cap {
            self.pos = 0;
            self.cap = 0;
        } else {
            self.buf.as_mut().copy_within(self.pos..self.cap, 0);
            self.cap -= self.pos;
            self.pos = 0;
        }
    }
}

impl<R: Read, B: AsRef<[u8]> + AsMut<[u8]>> Read for EnsuredBufReader<R, B> {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        let n = self.fill_buf()?.read(buf)?;
        self.consume(n);
        Ok(n)
    }
}

impl<R: Read, B: AsRef<[u8]> + AsMut<[u8]>> BufRead for EnsuredBufReader<R, B> {
    fn fill_buf(&mut self) -> io::Result<&[u8]> {
        self.fill_buf_to_expected_size(self.ensured_size)
    }

    fn consume(&mut self, amt: usize) {
        assert!(
            amt <= self.current_bytes(),
            "the amt must be <= the number of bytes in the buffer returned by fill_buf."
        );
        self.pos += amt;
    }
}

impl<R, B> fmt::Debug for EnsuredBufReader<R, B>
where
    R: Read + fmt::Debug,
    B: AsRef<[u8]> + AsMut<[u8]>,
{
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt.debug_struct("EnsuredBufReader")
            .field("reader", &self.inner)
            .field(
                "buffer",
                &format_args!("{}/{}", self.cap - self.pos, self.buf.as_ref().len()),
            )
            .finish()
    }
}

/// An error type may be returned from [`.fill_buf_to_expected_size()`](struct.EnsuredBufReader.html#method.fill_buf_to_expected_size).
#[derive(Debug, Clone, Copy)]
pub struct ExpectedSizeTooLargeError();

impl fmt::Display for ExpectedSizeTooLargeError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "internal buffer is too small.")
    }
}

impl error::Error for ExpectedSizeTooLargeError {}
